overlord/snapstate/snapmgr.go

```diff
@@ -861,10 +882,14 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 	cand := ss.SideInfo
 	m.backend.Candidate(cand)
 
-	hadCandidate := true
-	if snapst.LastIndex(cand.Revision) < 0 {
+	oldCandidateIndex := snapst.LastIndex(cand.Revision)
+
+	if oldCandidateIndex < 0 {
 		snapst.Sequence = append(snapst.Sequence, cand)
-		hadCandidate = false
+	} else if !ss.Flags.Revert() {
+		// remove the old candidate from the sequence, add it at the end
+		copy(snapst.Sequence[oldCandidateIndex:len(snapst.Sequence)-1], snapst.Sequence[oldCandidateIndex+1:])
```

Comment:
“it's hard for the computer to do, so it should be hard for you to write!” or something